File: BinarySearch/Sources/BinarySearch/BinarySearch.swift

// 	BinarySearch.swift
// 	Created by Ardalan Samimi on 2018-06-01
//
// 	Binary search is a search algorithm for finding
// 	the index of a specific key in a sorted array,
// 	where you repeatedly divide the list in half.
//
// 	Worst-case complexity: O(log n)
//

func binarySearch<T: Comparable>(for key: T, in array: [T]) -> Int? {
	// return binarySearch(array, key: key)
	guard array.count > 0 else { return nil }

	var minimum = 0
	var maximum = array.count - 1

	while maximum >= minimum {
		let middle = (maximum + minimum) / 2

		if key > array[middle] {
			minimum = middle + 1
		} else if key < array[middle] {
			maximum = middle - 1
		} else {
			return middle
		}
	}

	return nil
}

public func binarySearch<T: Comparable>(_ a: [T], key: T) -> Int? {
    var lowerBound = 0
    var upperBound = a.count
    while lowerBound < upperBound {
        let midIndex = lowerBound + (upperBound - lowerBound) / 2
        if a[midIndex] == key {
            return midIndex
        } else if a[midIndex] < key {
            lowerBound = midIndex + 1
        } else {
            upperBound = midIndex
        }
    }
    return nil
}

File: BinarySearch/Java/Sources/BinarySearch/BinarySearch.java
//
//  BinarySearch.java
//  BinarySearch
//
//  Created by Ardalan Samimi on 2019-03-29.
//
package BinarySearch;

public class BinarySearch {

		public static int search(int[] array, int element) {
			return BinarySearch.search(array, element, 0, array.length);
		}

		private static int search(int[] array, int element, int start, int end) {
			int size   = array.length;
			int middle = (start + end) / 2;

			if (start > end || middle >= size) return -1;

			if (element == array[middle]) {
				return middle;
			} else if (element < array[middle]) {
				return BinarySearch.search(array, element, start, middle - 1);
			} else if (element > array[middle]) {
				return BinarySearch.search(array, element, middle + 1, end);
			}

			return -1;
		}

}
